### PR TITLE
NDEV-2059: Improve performance of get_account_rooted_slot, get_branch_slots, get_sol_sig_rooted_slot

### PR DESCRIPTION
**Pull request by Deniskore**
Since most queries for get_account_rooted_slot result in either the last rooted_slot or adjacent slots, the first query should execute on a single day (input slot - 216000 slots)

This removes the fullscan on the update_slot table for a lot of requests.

Original with full scan:
```
SELECT DISTINCT slot
FROM events.update_account_distributed
WHERE (pubkey = [238, 205, 99, 85, 74, 252, 237, 43, 151, 11, 212, 155, 237, 78, 226, 30, 144, 127, 0, 158, 237, 154, 182, 209, 81, 245, 77, 5, 91, 181, 49, 210]) AND (slot <= 240920250) AND (slot IN (
    SELECT slot
    FROM events.update_slot
    WHERE status = 'Rooted'
    ORDER BY slot DESC
))
ORDER BY slot DESC
LIMIT 1

Query id: 90185cca-e051-4af4-9971-ed495332bcfc

┌──────slot─┐
│ 240920250 │
└───────────┘

1 row in set. Elapsed: 4.245 sec. Processed 45.31 million rows, 397.17 MB (10.67 million rows/s., 93.57 MB/s.)
```

New one:
```
SELECT DISTINCT slot
FROM events.update_account_distributed
WHERE (pubkey = [238, 205, 99, 85, 74, 252, 237, 43, 151, 11, 212, 155, 237, 78, 226, 30, 144, 127, 0, 158, 237, 154, 182, 209, 81, 245, 77, 5, 91, 181, 49, 210]) AND (slot <= 240920250) AND (slot IN (
    SELECT slot
    FROM events.update_slot
    WHERE (status = 'Rooted') AND (slot >= GREATEST(0, 240920250 - 216000))
))
ORDER BY slot DESC
LIMIT 1

Query id: ef49ce36-0739-4a6c-ba72-38367fdaf63d

┌──────slot─┐
│ 240920250 │
└───────────┘

1 row in set. Elapsed: 0.092 sec. Processed 933.37 thousand rows, 8.64 MB (10.18 million rows/s., 94.23 MB/s.)
```

### CODE REVIEW

**Review by dk-neon — APPROVED** [review]
LGTM

File: evm_loader/lib/src/types/tracer_ch_db.rs [hunk 1]
```diff
@@ -244,22 +262,33 @@ impl ClickHouseDb {
...
-                .query(query)
+                .query(query_last_day)
                 .bind(key)
                 .bind(slot)
                 .fetch_one::<u64>()
                 .await,
-        )?;
+        );
+
+        if slot_opt.is_err() {
```
Comment (dk-neon) [inline evm_loader/lib/src/types/tracer_ch_db.rs:274]: If no records were returned, this will be false.

**Review by dk-neon — CHANGES_REQUESTED** [review]

File: evm_loader/lib/src/types/tracer_ch_db.rs (hunk 1 above)
Comment (Deniskore) [reply evm_loader/lib/src/types/tracer_ch_db.rs:274]: Yep, but the code of row_opt a little bit misleading.
I'll remove the usage of row_opt.

```Rust
    fn row_opt<T>(result: clickhouse::error::Result<T>) -> clickhouse::error::Result<Option<T>> {
        match result {
            Ok(row) => Ok(Some(row)),
            Err(clickhouse::error::Error::RowNotFound) => Ok(None),
            Err(e) => Err(e),
        }
    }
```

File: evm_loader/lib/src/types/tracer_ch_db.rs [hunk 2]
```diff
@@ -244,22 +262,31 @@ impl ClickHouseDb {
...
 
         let execution_time = Instant::now().duration_since(time_start);
         info!(
             "get_account_rooted_slot {{ key: {key}, slot: {slot} }} sql(1) returned {slot_opt:?}, time: {} sec",
             execution_time.as_secs_f64(),
         );
 
-        Ok(slot_opt)
+        Ok(slot_opt.ok())
     }
```
Comment (dk-neon) [inline evm_loader/lib/src/types/tracer_ch_db.rs:290]: Current solution will return None on any CH error. Such behaviour is incorrect. It's expected to raise such errors and return Ok(None) if no rows were found.

**Review by dk-neon — CHANGES_REQUESTED** [review]

**Comment by Deniskore** [comment]
I've introduced a new table called rooted_slots in pull request https://github.com/neonlabsorg/geyser-neon-filter/pull/45, which focuses exclusively on rooted slots. Leveraging this table has led to significant performance improvements in queries.

File: evm_loader/lib/src/types/tracer_ch_db.rs [hunk 3]
```diff
@@ -229,18 +244,19 @@ impl ClickHouseDb {
...
-                AND slot <= ?
-                AND slot IN (
-                    SELECT slot
-                    FROM events.update_slot
-                    WHERE status = 'Rooted'
-                )
-            ORDER BY slot DESC
-            LIMIT 1
+        SELECT DISTINCT uad.slot
+        FROM events.update_account_distributed AS uad
```
Comment (dk-neon) [inline evm_loader/lib/src/types/tracer_ch_db.rs:250]: We don't need alias in this query

File: evm_loader/lib/src/types/tracer_ch_db.rs [hunk 4]
```diff
@@ -229,18 +244,19 @@ impl ClickHouseDb {
...
+        WHERE uad.pubkey = ?
+          AND uad.slot <= ?
+          AND (
+            SELECT COUNT(slot)
```
Comment (dk-neon) [inline evm_loader/lib/src/types/tracer_ch_db.rs:254]: This subquery isn't connected with the outer one. Can we extract this subquery to separated one and do the second query only on success?

**Review by dk-neon — CHANGES_REQUESTED** [review]

**Review by dk-neon — APPROVED** [review]

**Review by stanislav-tkach — APPROVED** [review]

**Review by andreisilviudragnea — APPROVED** [review]

File: evm_loader/lib/src/types/tracer_ch_db.rs [hunk 5]
```diff
@@ -49,6 +49,22 @@ pub struct SlotParent {
     pub status: u8,
 }
 
+#[derive(Debug, Row, serde::Deserialize, Clone)]
+pub struct SlotParentRooted {
```
[the remark this answers is not in the document]
Comment (Deniskore) [reply evm_loader/lib/src/types/tracer_ch_db.rs:53]: Let's stick with the current one. I'd like the name to begin similarly to 'SlotParent.' When one name is nested within another, it tends to be more perceptually straightforward and there is no need to run through the code if you want to change something.

**Pull request merged**